src/db.c

```diff
@@ -311,10 +328,20 @@ static void dbSetValue(serverDb *db, robj *key, robj *val, int overwrite, dictEn
         /* We want to try to unblock any module clients or clients using a blocking XREADGROUP */
         signalDeletedKeyAsReady(db, key, old->type);
         decrRefCount(old);
-        /* Because of RM_StringDMA, old may be changed, so we need get old again */
-        old = dictGetVal(de);
+        /* Because of VM_StringDMA, old may be changed, so we need get old again */
+        old = *oldref;
+    }
+    /* Replace the old value at its location in the key space. */
+    long long expire = objectGetExpire(old);
+    valkey *new = objectSetKeyAndExpire(val, key->ptr, expire);
```

Comment:
Consider reusing the old object and only freeing `obj->ptr`, then setting the `ptr` to point to the new value when both the old and new objects are not embstr. This would be similar to how we currently reuse dictEntry on overwrite.